html.go

```diff
@@ -819,55 +818,71 @@ func (r *HTMLRenderer) RenderNode(w io.Writer, node *Node, entering bool) WalkSt
 	return GoToNext
 }
 
-func (r *HTMLRenderer) writeDocumentHeader(w *bytes.Buffer) {
+// RenderHeader writes HTML document preamble and TOC if requested.
+func (r *HTMLRenderer) RenderHeader(w io.Writer, ast *Node) {
+	r.writeDocumentHeader(w)
+	if r.Flags&TOC != 0 {
+		r.writeTOC(w, ast)
+	}
+}
+
+// RenderFooter writes HTML document footer.
+func (r *HTMLRenderer) RenderFooter(w io.Writer, ast *Node) {
+	if r.Flags&CompletePage == 0 {
+		return
+	}
+	w.Write([]byte("\n</body>\n</html>\n"))
+}
+
+func (r *HTMLRenderer) writeDocumentHeader(w io.Writer) {
 	if r.Flags&CompletePage == 0 {
 		return
 	}
 	ending := ""
 	if r.Flags&UseXHTML != 0 {
-		w.WriteString("<!DOCTYPE html PUBLIC \"-//W3C//DTD XHTML 1.0 Transitional//EN\" ")
-		w.WriteString("\"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd\">\n")
-		w.WriteString("<html xmlns=\"http://www.w3.org/1999/xhtml\">\n")
+		w.Write([]byte("<!DOCTYPE html PUBLIC \"-//W3C//DTD XHTML 1.0 Transitional//EN\" "))
+		w.Write([]byte("\"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd\">\n"))
+		w.Write([]byte("<html xmlns=\"http://www.w3.org/1999/xhtml\">\n"))
 		ending = " /"
 	} else {
-		w.WriteString("<!DOCTYPE html>\n")
-		w.WriteString("<html>\n")
+		w.Write([]byte("<!DOCTYPE html>\n"))
+		w.Write([]byte("<html>\n"))
 	}
-	w.WriteString("<head>\n")
-	w.WriteString("  <title>")
+	w.Write([]byte("<head>\n"))
+	w.Write([]byte("  <title>"))
 	if r.Flags&Smartypants != 0 {
 		r.sr.Process(w, []byte(r.Title))
 	} else {
 		escapeHTML(w, []byte(r.Title))
 	}
-	w.WriteString("</title>\n")
-	w.WriteString("  <meta name=\"GENERATOR\" content=\"Blackfriday Markdown Processor v")
-	w.WriteString(Version)
-	w.WriteString("\"")
-	w.WriteString(ending)
-	w.WriteString(">\n")
-	w.WriteString("  <meta charset=\"utf-8\"")
-	w.WriteString(ending)
-	w.WriteString(">\n")
+	w.Write([]byte("</title>\n"))
+	w.Write([]byte("  <meta name=\"GENERATOR\" content=\"Blackfriday Markdown Processor v"))
+	w.Write([]byte(Version))
+	w.Write([]byte("\""))
+	w.Write([]byte(ending))
+	w.Write([]byte(">\n"))
+	w.Write([]byte("  <meta charset=\"utf-8\""))
+	w.Write([]byte(ending))
+	w.Write([]byte(">\n"))
 	if r.CSS != "" {
-		w.WriteString("  <link rel=\"stylesheet\" type=\"text/css\" href=\"")
+		w.Write([]byte("  <link rel=\"stylesheet\" type=\"text/css\" href=\""))
 		escapeHTML(w, []byte(r.CSS))
-		w.WriteString("\"")
-		w.WriteString(ending)
-		w.WriteString(">\n")
+		w.Write([]byte("\""))
+		w.Write([]byte(ending))
+		w.Write([]byte(">\n"))
 	}
 	if r.Icon != "" {
-		w.WriteString("  <link rel=\"icon\" type=\"image/x-icon\" href=\"")
+		w.Write([]byte("  <link rel=\"icon\" type=\"image/x-icon\" href=\""))
 		escapeHTML(w, []byte(r.Icon))
-		w.WriteString("\"")
-		w.WriteString(ending)
-		w.WriteString(">\n")
+		w.Write([]byte("\""))
+		w.Write([]byte(ending))
+		w.Write([]byte(">\n"))
 	}
-	w.WriteString("</head>\n")
-	w.WriteString("<body>\n\n")
+	w.Write([]byte("</head>\n"))
+	w.Write([]byte("<body>\n\n"))
```

Comment:
I see you had to change these calls from `WriteString` to `Write` because the parameter is now `io.Writer` instead of the previous `*bytes.Buffer`.

Question, have you considered using [`io.WriteString`](https://godoc.org/io#WriteString) instead? E.g.:

```diff
-w.WriteString("<body>\n\n")
+io.WriteString(w, "<body>\n\n")
```

How do you think it compares?